java-support/src/main/java/io/cloudstate/javasupport/KeyValue.java

```diff
@@ -0,0 +1,243 @@
+                  k.name(),
+                  com.google.protobuf.ByteString.copyFrom(
+                      ((Key<Object>) k).writer().apply(v).asByteBuffer()));
+            }
+          });
+      return builder.build();
+    }
+
+    final KVModification toProtoModification() {
+      final KVModification.Builder builder = KVModification.newBuilder();
+      updated.forEach(
+          (k, v) -> {
+            // Skip those which remain as unparsed, as they have not been changed
+            if (!unparsed.containsKey(k.name())) {
+              builder.putUpdatedEntries(
+                  k.name(),
+                  com.google.protobuf.ByteString.copyFrom(
+                      ((Key<Object>) k).writer().apply(v).asByteBuffer()));
+            }
+          });
+      removed.forEach(builder::addRemovedKeys);
+      return builder.build();
+    }
+
+    final void resetTo(KVEntity entityState) {
+      unparsed.clear();
+      updated.clear();
+      removed.clear();
+      entityState
+          .getEntriesMap()
+          .forEach(
+              (k, v) ->
+                  unparsed.put(
+                      k,
+                      v.isEmpty()
+                          ? ByteString.empty()
+                          : akka.util.ByteString.fromArrayUnsafe(v.toByteArray())));
+    }
+
+    final void applyModification(KVModificationOrBuilder modification) {
+      // Apply new modifications to the base unparsed values
+      modification
+          .getUpdatedEntriesMap()
+          .forEach(
+              (k, v) ->
+                  unparsed.put(
+                      k,
+                      v.isEmpty()
+                          ? ByteString.empty()
+                          : akka.util.ByteString.fromArrayUnsafe(v.toByteArray())));
+
+      modification.getRemovedKeysList().forEach(unparsed::remove);
+    }
+  }
+
+  @EventSourcedEntity
+  public abstract static class KeyValueEntity {
+    private final Map state = new Map();
+
+    protected Map state() {
+      return state;
+    }
+
+    @Snapshot
+    public KVEntityOrBuilder snapshot() {
+      return state.toProto();
+    }
+
+    @SnapshotHandler
+    public void handleSnapshot(final KVEntity entityState) {
+      state.resetTo(entityState);
+    }
+
+    @EventHandler
+    public void kVModification(final KVModification modification) {
+      state.applyModification(modification);
+    }
+  }
+
+  public static class YourCommandType {}
+
+  // We'll most likely want to add a KeyValueEntity annotation instead
+  public static class YourClass extends KeyValueEntity {
+    private static final Key<String> name =
+        keyOf("name", ByteString::utf8String, ByteString::fromString);
+
+    // We'll most likely want to add a specific CommandContext for KeyValueEntity
+    // so you'd instead of state().set(…) would do ctx.setState(…), and then it'd be automatically
+    // flushing ctx.emit(state.toProtoModification()) if there are any changes
+    @CommandHandler
```

Comment:
@jroper What do you think about the comments above, James? Do you think this is something we could put together rather quickly?

Reply:
I like that idea.

Reply:
@jroper ping :)

Reply:
@viktorklang Any progress here? This seems to be stopped for a long time

Reply:
@sleipnir No, unfortunately not. I guess the question is: Should we go for an API-only approach or support KV natively in the proxy as a different way of storing data.

Reply:
Hi @viktorklang
I am suspicious to answer because, I almost always end up voting for the proxy. But I think delegating to the proxy is the most elegant and transparent approach. Even taking advantage of persistence with Cassandra, Postgres, or key value native applications like Redis, Riak, Hazelcast, Infinispan, or anything like that. It's the kind of thing that would be worth exposing user interfaces to implement its own persistence mechanism (managed by the proxy of course)

Reply:
Hi @sleipnir,
I'm inclined to agree—I think it is much more valuable for the proxy to understand the specific state model and be able to better accommodate for its access patterns and optimization opportunities.

Reply:
@viktorklang @sleipnir do we have consensus how we would go forward with this?
@pvlugter @jroper do you have this on your roadmap?

Reply:
I think someone is going to have to get their hands dirty here and try to create something to start interacting better. I haven't seen any recent progress here @marcellanz . @ralphlaude  has an implementation proposal but I don't know how much he has evolved in that.

Reply:
@sleipnir i agree with you, someone have to get their hands dirty here :). I have thinking a bit about the a possible implementation for the native support and also the grpc protocol. But i didn't get that far. I am looking forward to get more involved here when other stuff going on are done :).

@marcellanz very good question and I am not sure there is a consensus here. If there is a consensus, I think the first step here would be to define the grpc protocol.

Reply:
The CRUD support will provide the underlying native KV support discussed in this thread, and @ralphlaude is making progress there. (Side note: not sure if "CRUD" entities is the best name, maybe we can improve the naming there too.) Until that's complete, language supports could also experiment with including a `setState` or `setValue` style API directly on event sourced entities, setting a snapshot value, if we want to try this out earlier.

Also agree that it would be good to try out some use cases, to see how useful a KV API like in this PR might be, since this would need to be added to all the language supports. This currently gives a more generic Map-like API in place of defining protobuf message objects directly (which can also be map-like as needed). The shopping cart samples also use a map for state — would that example be improved with a generic key-value API, compared with the types it uses now for event sourcing? In these kinds of examples, would just setting a map as the value (and using CRUD support) already be clean enough?

Entity services themselves provide a KV-style store, keyed by entity id, and can be backed by event sourced, or the upcoming CRUD for a simple set-value approach, or using CRDTs for distributed, in-memory, eventually consistent storage. Within an entity the state can be a map (be key-value-based) like in this PR.

And then there's the concept of a KV store that's available across Entity or Action instances — like the Cloudflare Workers KV, which is an eventually-consistent LWW (last writer wins) key-value store distributed across the worker nodes. For Cloudstate, the pattern that we've demonstrated that's similar to this is having multiple entity types. Like the "hot items" CRDT entity used alongside the shopping cart. We have forwarding and side effects to combine entities. Maybe we should look at supporting an "ask pattern", an async get on another entity type, where values can be fetched from other entity types and then continue processing with the responses.

Reply:
@jroper and I discussed this further. Some things we discussed:

We don't currently see clear use cases for a key-value API like this in the language supports. Especially when compared with simply using an object or map as the value (with CRUD or event sourcing support). For persisting deltas/changes only, there's the event sourcing already, and having an additional KV API increases the surface area that language supports need to implement.

For CRUD support naming, we're suggesting to refer to this as a value-based state model, and it could be the default state model for entities. From the user perspective, we could simply call these `Entities`. And then there are `Event Sourced Entities` and `Replicated Entities` (CRDTs). Within code or discussions, we could use value entities or value-based entities or persisted value entities as needed. And for services, entities themselves provide key-value storage (with the entity key and its value), which can be backed by simple key-value storage, event sourcing, or in-memory replication.

I think the main thing to do here is to complete the basic entities with value-based state model (CRUD support). In terms of roadmap, we also discussed other forward-looking things, like replicated and active-active event-sourcing, durable actions (as in long-lasting, that can fetch asynchronously, implement workflows), durable CRDTs, and product CRDTs.

So suggesting that we close this PR and put the idea of KV APIs in language supports on hold, and focus on finishing the value entities (CRUD support) and making those the default starting point. Interested to hear more thoughts on this, and it can be a topic for the next contributors' call.

Reply:
I agree that this is the best path forward.

Reply:
Having a good "CRUD" solution might obviate the need for a KV solution, especially if the proxy can diff.